src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -201,101 +193,51 @@ internal bool TryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSeriali
                         ref reader);
                 }
 
-                if (options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.Preserve &&
-                    CanBePolymorphic && value is JsonElement element)
-                {
-                    // Edge case where we want to lookup for a reference when parsing into typeof(object)
-                    // instead of return `value` as a JsonElement.
-                    Debug.Assert(TypeToConvert == typeof(object));
-
-                    if (JsonSerializer.TryGetReferenceFromJsonElement(ref state, element, out object? referenceValue))
-                    {
-                        value = (T?)referenceValue;
-                    }
-                }
-
                 return true;
             }
 
+            Debug.Assert(IsInternalConverter);
+            bool isContinuation = state.IsContinuation;
             bool success;
 
-            // Remember if we were a continuation here since Push() may affect IsContinuation.
-            bool wasContinuation = state.IsContinuation;
-
 #if DEBUG
             // DEBUG: ensure push/pop operations preserve stack integrity
             JsonTypeInfo originalJsonTypeInfo = state.Current.JsonTypeInfo;
 #endif
             state.Push();
             Debug.Assert(TypeToConvert == state.Current.JsonTypeInfo.Type);
 
-#if !DEBUG
+#if DEBUG
             // For performance, only perform validation on internal converters on debug builds.
```

Comment:
It appears we are now not validating custom converters (non-internal) in release build (the previous code always validated custom converters).

Reply:
It turns out that the `IsInternalConverter` check was redundant here, since it's always true here. I've added an assert in line 199 that checks for that.